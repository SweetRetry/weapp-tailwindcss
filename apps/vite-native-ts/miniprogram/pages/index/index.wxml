<view class="p-4">
  <IceButton class="bg-[#d72929]" my-class="bg-[orange]"></IceButton>
</view>