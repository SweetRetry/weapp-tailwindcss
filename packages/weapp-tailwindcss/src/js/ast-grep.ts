import { logger } from '../logger'

export async function getAstGrep() {
  try {
    const { js } = await import('@ast-grep/napi')
    return js
  }
  catch (error) {
    logger.warn('请先安装 `@ast-grep/napi` , 安装完成后再尝试运行！')
    throw error
  }
}
