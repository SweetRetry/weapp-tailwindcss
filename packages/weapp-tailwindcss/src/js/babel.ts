import type { NodePath } from '@babel/traverse'
import type { Node } from '@babel/types'

export function isEvalPath(p: NodePath<Node>) {
  if (p.isCallExpression()) {
    const calleePath = p.get('callee')
    return calleePath.isIdentifier() && calleePath.node.name === 'eval'
  }
  return false
}
