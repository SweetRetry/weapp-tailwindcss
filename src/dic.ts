export const SYMBOL_TABLE = {
  BACKQUOTE: '`',
  TILDE: '~',
  EXCLAM: '!',
  AT: '@',
  NUMBERSIGN: '#',
  DOLLAR: '$',
  PERCENT: '%',
  CARET: '^',
  AMPERSAND: '&',
  ASTERISK: '*',
  PARENLEFT: '(',
  PARENRIGHT: ')',
  MINUS: '-',
  UNDERSCORE: '_',
  EQUAL: '=',
  PLUS: '+',
  BRACKETLEFT: '[',
  BRACELEFT: '{',
  BRACKETRIGHT: ']',
  BRACERIGHT: '}',
  SEMICOLON: ';',
  COLON: ':',
  QUOTE: "'",
  DOUBLEQUOTE: '"',
  BACKSLASH: '\\',
  BAR: '|',
  COMMA: ',',
  LESS: '<',
  PERIOD: '.',
  GREATER: '>',
  SLASH: '/',
  QUESTION: '?',
  SPACE: '',
  DOT: '.',
  HASH: '#'
} as const

export const MappingChars2String = {
  [SYMBOL_TABLE.BRACKETLEFT]: '_l_',
  [SYMBOL_TABLE.BRACKETRIGHT]: '_r_',
  [SYMBOL_TABLE.PARENLEFT]: '_p_',
  [SYMBOL_TABLE.PARENRIGHT]: '_q_',
  [SYMBOL_TABLE.HASH]: '_h_',
  [SYMBOL_TABLE.EXCLAM]: '_i_',
  [SYMBOL_TABLE.SLASH]: '_div_',
  [SYMBOL_TABLE.DOT]: '_dot_',
  [SYMBOL_TABLE.COLON]: '_c_',
  [SYMBOL_TABLE.PERCENT]: '_pct_',
  [SYMBOL_TABLE.COMMA]: '_d_',
  [SYMBOL_TABLE.QUOTE]: '_y_'
} as const
