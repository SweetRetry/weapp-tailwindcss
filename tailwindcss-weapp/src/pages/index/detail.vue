<script setup lang="ts">
import BaseLayout from '@/components/BaseLayout.vue'
import Fab from '@/components/FloatButton.vue'
import Navbar from '@/components/Navbar.vue'
import { useSystemStore } from '@/stores'
import camelCase from 'lodash/camelCase'
import redent from 'redent'

const store = useSystemStore()

interface TableItem {
  Class: string
  Properties: string[]
}

const listData = ref<TableItem[]>([])
const tipsContent = `${['长按行可以复制当前行的值']
  .map((x, idx) => {
    return `${idx + 1}. ${x}`
  })
  .join('\n')}`

const plugin = ref('')
const title = ref('')
const tipVisible = ref(false)
const header = [
  {
    title: 'Class',
    key: 'Class',
  },
  {
    title: 'Properties',
    key: 'Properties',
  },
]

const currentUrl = computed(() => {
  return `/pages/index/detail?t=${title.value}`
})

function showTips() {
  tipVisible.value = true
}

function longpressCopy(item: TableItem) {
  const copyValue = [
    `${item.Class} {`,
    `  ${item.Properties.join('\n  ')}`,
    '}',
  ].join('\n')

  uni.setClipboardData({
    data: copyValue,
  })
}

onLoad((params) => {
  title.value = params?.t as string
  plugin.value = camelCase(title.value)
  switch (plugin.value) {
    case 'topRightBottomLeft': {
      plugin.value = 'inset'
      break
    }
    case 'screenReaders': {
      plugin.value = 'accessibility'
      break
    }
    default: {
      break
    }
  }
  if (plugin.value === 'animation') {
    listData.value = [
      {
        Class: 'animate-none',
        Properties: ['animation: none;'],
      },
      {
        Class: 'animate-spin',
        Properties: [
          redent(`
              animation: spin 1s linear infinite;\n
              @keyframes spin {
                from {
                  transform: rotate(0deg);
                }
                to {
                  transform: rotate(360deg);
                }
              }
            `).trim(),
        ],
      },
      {
        Class: 'animate-ping',
        Properties: [
          redent(`
              animation: ping 1s cubic-bezier(0, 0, 0.2, 1) infinite;\n
              @keyframes ping {
                75%, 100% {
                  transform: scale(2);
                  opacity: 0;
                }
              }
            `).trim(),
        ],
      },
      {
        Class: 'animate-pulse',
        Properties: [
          redent(`
              animation: pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;\n
              @keyframes pulse {
                0%, 100% {
                  opacity: 1;
                }
                50% {
                  opacity: .5;
                }
              }
            `).trim(),
        ],
      },
      {
        Class: 'animate-bounce',
        Properties: [
          redent(`
              animation: bounce 1s infinite;\n
              @keyframes bounce {
                0%, 100% {
                  transform: translateY(-25%);
                  animation-timing-function: cubic-bezier(0.8, 0, 1, 1);
                }
                50% {
                  transform: translateY(0);
                  animation-timing-function: cubic-bezier(0, 0, 0.2, 1);
                }
              }
            `).trim(),
        ],
      },
    ]
  }
  else {
    listData.value = Object.entries(store.getPluginsById(plugin.value)).map(
      (x) => {
        return {
          Class: x[0],
          // @ts-ignore
          Properties: Object.entries(x[1]).map(([key, value]) => {
            return `${key}: ${value};`
          }),
        }
      },
    )
  }

  uni.setNavigationBarTitle({
    title: title.value,
  })
})

onShareAppMessage(() => {
  return {
    title: plugin.value,
    path: currentUrl.value,
  }
})

onShareTimeline(() => {
  return {
    title: plugin.value,
    path: currentUrl.value,
  }
})
</script>

<template>
  <BaseLayout>
    <Navbar auto-back>
      <template #center>
        {{ title }}
      </template>
    </Navbar>
    <view class="px-4">
      <view class="tw-table">
        <view class="thead sticky top-0 z-50 bg-white transition-colors duration-500 dark:bg-slate-900">
          <view class="tr">
            <view class="w-4/12">
              <view class="flex py-2 pr-2 text-sm font-semibold text-slate-700 dark:text-slate-300">
                <text class="mr-2">
                  {{ header[0].title }}
                </text>
                <u-icon name="question-circle" @click="showTips" />
              </view>
            </view>
            <view class="w-8/12">
              <view class="py-2 pl-2 text-sm font-semibold text-slate-700 dark:text-slate-300">
                {{
                  header[1].title
                }}
              </view>
            </view>
          </view>
        </view>
        <view class="tbody">
          <view
            v-for="(item, index) in listData" :key="index" class="tr" hover-class="hover-gray-100"
            @longpress="longpressCopy(item)"
          >
            <template v-for="(obj, idx) in header" :key="idx">
              <view :class="idx === 0 ? 'w-4/12' : 'w-8/12'">
                <template v-if="Array.isArray(item[obj.key])">
                  <view class="whitespace-pre-wrap py-2 pl-2  text-xs text-indigo-600 dark:text-indigo-300">
                    <view v-for="(r, i) in item[obj.key]" :key="i">
                      {{ r }}
                    </view>
                  </view>
                </template>
                <template v-else>
                  <view
                    class="
                    whitespace-pre-wrap
                    py-2
                    pr-2 text-xs font-medium
                    text-sky-500
                    dark:text-sky-400
                  "
                  >
                    {{ item[obj.key] }}
                  </view>
                </template>
              </view>
            </template>
          </view>
        </view>
        <!-- <view></view> -->
      </view>

      <u-modal :show="tipVisible" title="操作提示" :content="tipsContent" @confirm="tipVisible = false" />
      <Fab store-key="detail-float-btn">
        <button
          class="
          u-reset-button
          pointer-events-auto
          flex
          size-10
          items-center
          justify-center
          rounded-full
          bg-white
          shadow-sm
          dark:bg-slate-700

        " open-type="share"
        >
          <i class="i-ri-wechat-fill text-xl text-[#07c160]" />
        </button>
      </Fab>
    </view>
  </BaseLayout>
</template>

<style lang="scss">
.tw-table {
  .thead {
    .tr {
      @apply border-b border-slate-200 dark:border-slate-400/20;
    }
  }

  .tbody {
    .tr {
      @apply border-b border-slate-100 dark:border-slate-400/10;
    }
  }

  .tr {
    @apply flex;

    .td {
    }
  }
}
</style>
