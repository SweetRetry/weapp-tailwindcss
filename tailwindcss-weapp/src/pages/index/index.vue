<script setup lang="ts">
import { devDependencies } from '../../../package.json'
import BaseLayout from '@/components/BaseLayout.vue'
import FloatButton from '@/components/FloatButton.vue'
import { useSystemStore } from '@/stores'

const version = devDependencies.tailwindcss.slice(1)

const store = useSystemStore()

function openSearchBox() {

}

function toggleTheme() {
  if (store.theme === 'dark') {
    store.setTheme('light')
  }
  else {
    store.setTheme('dark')
  }
}

function go2ThemeDemo() {
  uni.navigateTo({
    url: '/pages/theme/index',
  })
}
</script>

<template>
  <BaseLayout>
    <up-navbar fixed bgColor="inherit" safeAreaInsetTop>
      <template #left>
        <view class="relative flex items-center">
          <view class="i-logos-tailwindcss-icon mr-2 h-[20.57px] w-[34px] shrink-0" />
          <view class="text-lg font-semibold">
            Tailwind CSS
          </view>
          <view
            class="relative bottom-0.5 ml-1 self-end rounded-full bg-sky-400/10 px-2 py-0.5 text-xs font-medium text-sky-600"
          >
            {{ version }}
          </view>
        </view>
      </template>
    </up-navbar>
    <FloatButton store-key="index-float-btn" :padding="[256, 32, 256, 32]">
      <view class="float-btn mb-3">
        <view class="i-mdi-flask-round-bottom-empty-outline text-sky-400 dark:text-sky-500" @click="go2ThemeDemo" />
      </view>
      <view class="float-btn mb-3" @click="toggleTheme">
        <view v-if="store.theme === 'dark'" class="i-mdi-white-balance-sunny text-sky-500" />
        <view v-else class="i-mdi-weather-night text-sky-400" />
      </view>
      <view class="float-btn" @click="openSearchBox">
        <view class="i-mdi-magnify text-sky-400 dark:text-sky-500" />
      </view>
    </FloatButton>
  </BaseLayout>
</template>
