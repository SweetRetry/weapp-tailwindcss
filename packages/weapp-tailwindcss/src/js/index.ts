import type { CreateJsHandlerOptions, IJsHandlerOptions } from '../types'
import { defuOverrideArray } from '../utils'
import { jsHandlerAsync } from './ast-grep'
import { jsHandler } from './babel'

export {
  jsHandler,
  jsHandlerAsync,
}

export function createJsHandler(options: CreateJsHandlerOptions) {
  const {
    mangleContext,
    arbitraryValues,
    escapeMap,
    jsPreserveClass,
    generateMap,
    jsAstTool,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
  } = options

  function handler(rawSource: string, classNameSet: Set<string>, options?: CreateJsHandlerOptions) {
    const opts = defuOverrideArray<IJsHandlerOptions, IJsHandlerOptions[]>(options as IJsHandlerOptions, {
      classNameSet,
      escapeMap,
      arbitraryValues,
      mangleContext,
      jsPreserveClass,
      generateMap,
      jsAstTool,
      babelParserOptions,
      ignoreCallExpressionIdentifiers,
      ignoreTaggedTemplateExpressionIdentifiers,
    })
    if (opts.jsAstTool === 'ast-grep') {
      return jsHandlerAsync(rawSource, opts)
    }
    return jsHandler(rawSource, opts)
  }

  function sync(rawSource: string, classNameSet: Set<string>, options?: CreateJsHandlerOptions) {
    const opts = defuOverrideArray<IJsHandlerOptions, IJsHandlerOptions[]>(options as IJsHandlerOptions, {
      classNameSet,
      escapeMap,
      arbitraryValues,
      mangleContext,
      jsPreserveClass,
      generateMap,
      jsAstTool,
      babelParserOptions,
      ignoreCallExpressionIdentifiers,
      ignoreTaggedTemplateExpressionIdentifiers,
    })
    return jsHandler(rawSource, opts)
  }

  handler.sync = sync

  return handler
}
