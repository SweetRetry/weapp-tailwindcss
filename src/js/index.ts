import { parse, traverse, generate } from '@/babel'
import type { TraverseOptions, IJsHandlerOptions } from '@/types'
import type { Node, StringLiteral, TemplateElement } from '@babel/types'
import { replaceWxml } from '@/wxml/shared'
import { escapeStringRegexp } from '@/reg'
import { splitCode } from '@/extractors/split'
import { isProd } from '@/env'

export function handleValue(str: string, node: StringLiteral | TemplateElement, options: IJsHandlerOptions) {
  const set = options.classNameSet
  const escapeEntries = options.escapeEntries
  const arr = splitCode(str) // .split(/\s/).filter((x) => x) // splitCode(n.value) // .split(/\s/).filter((x) => x)
  let rawStr = str
  for (let i = 0; i < arr.length; i++) {
    const v = arr[i]
    if (set.has(v)) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag = node.leadingComments.findIndex((x) => x.value.includes('weapp-tw') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        rawStr = rawStr.replace(
          new RegExp(escapeStringRegexp(v), 'g'),
          replaceWxml(v, {
            escapeEntries
          })
        )
      }
    }
  }
  return rawStr
}

export function jsHandler(rawSource: string, options: IJsHandlerOptions) {
  const ast = parse(rawSource)
  // 这样搞会把原先所有的 children 含有相关的 也都转义了
  const topt: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        const n = p.node
        n.value = handleValue(n.value, n, options)
      }
      // exit(p) {}
    },
    TemplateElement: {
      enter(p) {
        const n = p.node
        n.value.raw = handleValue(n.value.raw, n, options)
      }
    },
    noScope: true
  }

  traverse(ast, topt)

  return generate(ast, {
    minified: options.minifiedJs ?? isProd()
  })
}

export function createjsHandler(options: Omit<IJsHandlerOptions, 'classNameSet'>) {
  return (rawSource: string, set: Set<string>) => {
    return jsHandler(rawSource, {
      classNameSet: set,
      escapeEntries: options.escapeEntries,
      minifiedJs: options.minifiedJs
    })
  }
}
