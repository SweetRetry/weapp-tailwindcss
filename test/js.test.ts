/* eslint-disable no-template-curly-in-string */
import { SimpleMappingChars2StringEntries } from '@/dic'
import { createjsHandler } from '@/js/index'
describe('jsHandler', () => {
  it('common case', () => {
    const h = createjsHandler({
      escapeEntries: SimpleMappingChars2StringEntries
    })
    const set: Set<string> = new Set()
    set.add('text-[12px]')
    set.add('flex')
    set.add('w-2.5')
    const code = h(`const n = 'text-[12px] flex bg-[red] w-2.5'`, set).code
    expect(code).toBe('const n = "text-_12px_ flex bg-[red] w-2d5";')
  })

  it('preserve space', () => {
    const h = createjsHandler({
      escapeEntries: SimpleMappingChars2StringEntries
    })
    const set: Set<string> = new Set()
    set.add('text-[12px]')
    set.add('flex')
    set.add('w-2.5')
    // const code = h("const n = 'text-[12px] flex' + '\n' + ' bg-[red] w-2.5'", set).code
    const code = h("const n = 'text-[12px] flex \\n bg-[red] w-2.5'", set).code
    expect(code).toBe('const n = "text-_12px_ flex \\n bg-[red] w-2d5";')
  })

  it('preserve space case2', () => {
    const h = createjsHandler({
      escapeEntries: SimpleMappingChars2StringEntries
    })
    const set: Set<string> = new Set()
    set.add('text-[12px]')
    set.add('flex')
    const code = h("const n = `text-[12px] \\n\\n  flex  \\n\\n  bg-[red] '`", set).code
    expect(code).toBe("const n = `text-_12px_ \\n\\n  flex  \\n\\n  bg-[red] '`;")
  })

  it('babel TemplateElement case', () => {
    const h = createjsHandler({
      escapeEntries: SimpleMappingChars2StringEntries
    })
    const set: Set<string> = new Set()
    set.add('text-[12px]')
    set.add('flex')
    set.add('bg-[red]')
    const code = h("const p = 'text-[12px]';const n = `${p} \\n\\n  flex  \\n\\n  bg-[red] '`", set).code
    expect(code).toBe('const p = "text-_12px_";\nconst n = `${p} \\n\\n  flex  \\n\\n  bg-_red_ \'`;')
  })

  it('mpx jit classNames', () => {
    const testCase = `data: {
      classNames: "text-[#123456] text-[50px] bg-[#fff]"
    }`
    const h = createjsHandler({
      escapeEntries: SimpleMappingChars2StringEntries
    })
    const set: Set<string> = new Set()
    set.add('text-[#123456]')
    set.add('bg-[#fff]')
    set.add('text-[50px]')
    const code = h(testCase, set).code
    expect(code).toMatchSnapshot()
  })
})
