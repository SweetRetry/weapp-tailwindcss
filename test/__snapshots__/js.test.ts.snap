// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsHandler img url case 1`] = `
"data: {
  classNames: "bg-_url_qhttpscssylnavdcomsassetssimagessvusdivider-graydwebpq__";
}"
`;

exports[`jsHandler mpx jit classNames 1`] = `
"data: {
  classNames: "text-_h123456_ text-_50px_ bg-_hfff_";
}"
`;
